snowflake/custom-recipes/python-custom-recipes/recipe.py

```diff
@@ -1,113 +1,173 @@
+if config_in["formatType"] != 'csv':
+    print("[-] Only a CSV format for the input DSS Dataset is supported (you used {}).".format(config_in["formatType"]))
     print("[-] Please adjust the format. Aborting")
     sys.exit("Format error (CSV needed)")
 
-project_key = config["projectKey"]
+project_key = config_in["projectKey"]
 
 # Actual path of the input file on S3
-bucket = config["params"]["bucket"]
-path = config["params"]["path"].replace("${projectKey}",config["projectKey"])
+bucket = config_in["params"]["bucket"]
+path = config_in["params"]["path"].replace("${projectKey}", project_key)
 full_path = "s3://{}{}".format(bucket, path)
 
+
 # Input file definition
-separator = config["formatParams"]["separator"]
-skip_rows = config["formatParams"]["skipRowsBeforeHeader"]
+separator = config_in["formatParams"]["separator"]
+skip_rows = config_in["formatParams"]["skipRowsBeforeHeader"]
 
 
 #------------------------------------------------------------------------------
 # OUTPUT DATASET SETTINGS
 #------------------------------------------------------------------------------
 
-# Output configuration
-config = out.get_location_info(sensitive_info=True)
 
 # Snowflake credentials & output table
-jdbc_url = config["info"]["connectionParams"]["jdbcurl"]
-components = urlparse.parse_qs(urlparse.urlparse(jdbc_url).query)
+jdbc_url_parsed = urlparse.urlparse(config_out["connectionParams"]["jdbcurl"])
+components = urlparse.parse_qs(jdbc_url_parsed.query)       
+
+sf_cnx_args = {
+    "user": components["user"][0],
+    "password": components["password"][0],
+    "warehouse": components["warehouse"][0],
+    "database": components["db"][0]
+}
+
+m = re.match(r'snowflake://(?P<account>[^.]+)(\.(?P<region>[\w\d\-]+))?\.snowflakecomputing.com', jdbc_url_parsed.path, re.IGNORECASE)
+
+sf_cnx_args["account"] = m.group("account")
+
+if m.group("region"):
+    sf_cnx_args["region"] = m.group("region")
+
+if "schema" in components:
+    sf_cnx_args["schema"] = components["schema"][0]
+    print("[+] Connect using Schema '{}'".format(sf_cnx_args["schema"]))
+
+if "role" in components:
+    sf_cnx_args["role"] = components["role"][0]
+    print("[+] Connect using Role '{}'".format(sf_cnx_args["role"]))
 
-sf_user      = components["user"][0]
-sf_password  = components["password"][0]
-sf_database  = components["db"][0]
-sf_schema    = components["schema"][0]
-sf_warehouse = components["warehouse"][0]
-sf_account   = urlparse.urlparse(jdbc_url).path.replace("snowflake://", "").replace(".snowflakecomputing.com", "")
 
-output_table = config["info"]["table"].replace("${projectKey}", project_key)
+print("[+] Connecting to SF DB '{database}' using Warehouse '{warehouse}', and User '{user}'.".format(**sf_cnx_args))
+
+output_table = config_out["table"].replace("${projectKey}", project_key)
+output_schema = config_out["schema"].replace("${projectKey}", project_key)
```

Comment:
there should be a confusion here. This will raise an error when the DSS dataset is not already created. And in any cases DSS defines a schema as a (column_name, type) dictionary, this is not the schema as understood in Snowflake as a collection table. We'd better have the snowflake SCHEMA as a parameter of the plugin instead, maybe with default being PUBLIC

Reply:
The default should start with the schema specified in the JDBC URL (if one is specified). In what cases is the output dataset not configured? I thought you can't get this far in a recipe without setting an output?

This is the hierarchy of decisions that I'd been imagining:

1. Output table explicitly specifies a schema
2. Output table doesn't specify a schema, or doesn't yet exist. Use the schema specified by the JDBC URL.
3. If the JDBC URL does not have a schema parameter (technically it's optional), allow a user to specify one

Implementation of that would be a bit weird. If there were a more strongly typed Snowflake connection (rather than just the generic JDBC), you could require the schema parameter (and encrypt the password!) right there.

Reply:
understood.
The case to take into account is just:
- no schema in jdbc url
- no schema in output table (because not configured)
- no schema provided by the user
=> should we raise an explicit error "schema mandatory" ? or use PUBLIC by default with WARNING.
What is your view on this ?

Reply:
@kguedec I'd probably raise an error. (I'm not certain my teammates, for example, would look for a warning before telling me it's "broken.")